### PR TITLE
REFACTOR-#4282: rename `filepath_or_buffer_md` variable name as it is misleading since no buffer is used in it

### LINKED ISSUE #4282: REFACTOR: `filepath_or_buffer_md` variable name is misleading as no buffer is used in it
Initial found: https://github.com/modin-project/modin/pull/3441#discussion_r816719214

Current behavior: if the buffer is associated with a file, then an attempt is made to get the file name and work on it (`Modin` path), otherwise the code will execute in `pandas`.

### PR DESCRIPTION
**Pull request by anmyachev**
Signed-off-by: Anatoly Myachev <anatoly.myachev@intel.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4282 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date 
- [x] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release

### CODE REVIEW

File: modin/core/io/text/text_file_dispatcher.py [hunk 1]
```diff
@@ -997,6 +997,10 @@ def _read(cls, filepath_or_buffer, **kwargs):
                 filepath_or_buffer, callback=cls.read_callback, **kwargs
             )
 
+        # We know for sure that the buffer is not used further in the code,
+        # renaming in order not to be misleading
+        filepath = filepath_or_buffer_md
```
Comment (dchigarev) [inline modin/core/io/text/text_file_dispatcher.py:1002]: if the buffer isn't used, why do we try to extract it a few lines above? (`cls.get_path_or_buffer`)
https://github.com/modin-project/modin/blob/97769988a6f19e4b76f34238c97bf159ee7626a5/modin/core/io/text/text_file_dispatcher.py#L960-L964
We don't use `filepath_or_buffer_md` as a buffer **nowhere** in this function, not only after this renaming line. 

BTW, why don't we use the variable as a buffer? What happens if we pass a buffer? If the case is not supported, maybe we should raise a NonImplementedError on it instead of silently ignoring the fact that `filepath_or_buffer_md` may contain a file buffer?

Comment (anmyachev) [reply modin/core/io/text/text_file_dispatcher.py:1002]: `cls.get_path_or_buffer` tries to extract the filename in order to continue working with the string. However, if this cannot be done, a buffer is returned, which will fall into the check of supported types (into `check_parameters_support` function) and further into pandas. So a buffer **can be used** in the function before `check_parameters_support`+`single_worker_read`.

Buffer data type isn't serializable - root cause.
We do not silently ignore,  this case is executed in pandas.

Comment (dchigarev) [reply modin/core/io/text/text_file_dispatcher.py:1002]: I'm concerned that both of the variable names would be available in the function's namespace pointing to the same object (`filepath` and `filepath_or_buffer_md`), leaving a developer a choice of which one to use. That could end up in even more confusion if both of the variables be present in the flow mixed together (if somebody mistakenly would use the old name).

Maybe it's okay, but I don't really like it that way, let's leave an honor to someone else to put the approval :)

**Review by RehanSD — CHANGES_REQUESTED** [review]
Why don't we just change line 958 from 
```python
filepath_or_buffer_md =
```
To
```python
filepath_md =
```
?
I noticed the variable isn't really used prior to your reassignment, so maybe it makes sense to just change the name, and have a comment above that explains that if it's not a filepath Modin doesn't use it, so either it's a filepath or it's going to pandas?

**Comment by anmyachev** [comment]
@RehanSD what do you think about new approach?

File: modin/core/io/text/text_file_dispatcher.py [hunk 2]
```diff
@@ -642,6 +642,11 @@ def check_parameters_support(
                 return False
         elif not cls.pathlib_or_pypath(filepath_or_buffer):
             return False
+        elif not (
+            hasattr(filepath_or_buffer, "name")
+            and cls.file_exists(filepath_or_buffer.name)
+        ):
```
Comment (RehanSD) [inline modin/core/io/text/text_file_dispatcher.py:648]: What does this do?

**Review by RehanSD — COMMENTED** [review]
I like the new approach! Confused by one thing, so left a question!

File: modin/core/io/text/text_file_dispatcher.py (hunk 2 above)
Comment (anmyachev) [reply modin/core/io/text/text_file_dispatcher.py]: @RehanSD changed the code a bit and added a comment. Ready for review

**Comment by anmyachev** [comment]
@RehanSD ready for review

**Review by RehanSD — DISMISSED** [review]
LGTM!

**Comment by RehanSD** [comment]
Down to merge if @vnlitvinov is down

File: modin/core/io/text/text_file_dispatcher.py [hunk 3]
```diff
@@ -642,6 +642,10 @@ def check_parameters_support(
                 filepath_or_buffer, read_kwargs.get("storage_options")
             ):
                 return False
+        elif hasattr(filepath_or_buffer, "name"):
+            # `hasattr` check is needed for the case when modin can speed up reading from buffer
+            if not cls.file_exists(filepath_or_buffer.name):
+                return False
```
Comment (vnlitvinov) [inline modin/core/io/text/text_file_dispatcher.py:648]: I wonder how do we support this particular speed up case - do we open it separately on workers?.. what if incoming `filepath_or_buffer` was sought to certain point before passing it in?..

Also this is unrelated to PR title I think.

Comment (anmyachev) [reply modin/core/io/text/text_file_dispatcher.py:648]: This case is not accelerated in any way, it just defaults to pandas.

Comment (vnlitvinov) [reply modin/core/io/text/text_file_dispatcher.py:648]: I'm speaking about the case when `filepath_or_buffer.name` _does_ exist, but the user has given us the buffer with non-zero starting position.

Comment (anmyachev) [reply modin/core/io/text/text_file_dispatcher.py:648]: oops :) this looks like a bug

**Review by vnlitvinov — CHANGES_REQUESTED** [review]
Pls do not introduce new bugs: https://github.com/modin-project/modin/pull/4283#pullrequestreview-1035686407 🙃

**Comment by anmyachev** [comment]
> Pls do not introduce new bugs: [#4283 (review)](https://github.com/modin-project/modin/pull/4283#pullrequestreview-1035686407) 🙃

This bug is present in the master, I just moved the code to a slightly different place.

I'll create a task for this in a while.

**Comment by anmyachev** [comment]
Blocked by #4680

File: modin/core/io/text/text_file_dispatcher.py [hunk 4]
```diff
@@ -1033,8 +1036,9 @@ def _read(cls, filepath_or_buffer, **kwargs):
             nrows=None,
             compression=compression_infered,
         )
+
```
Comment (anmyachev) [inline modin/core/io/text/text_file_dispatcher.py:1039]:
```suggestion
```
